pyomo/contrib/mindtpy/outer_approximation.py

```diff
@@ -0,0 +1,445 @@
+
+# -*- coding: utf-8 -*-
+
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright (c) 2008-2022
+#  National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+import logging
+from pyomo.contrib.gdpopt.util import (time_code, lower_logger_level_to, copy_var_list_values)
+from pyomo.contrib.mindtpy.util import set_up_logger, setup_results_object, add_var_bound, calc_jacobians
+from pyomo.core import TransformationFactory, maximize, Objective, ConstraintList
+from pyomo.opt import SolverFactory
+from pyomo.contrib.mindtpy.config_options import _get_MindtPy_OA_config
+from pyomo.contrib.mindtpy.algorithm_base_class import _MindtPyAlgorithm
+from pyomo.contrib.mindtpy.feasibility_pump_new import MindtPy_FP_Solver
+from pyomo.contrib.mindtpy.util import get_integer_solution, copy_var_list_values_from_solution_pool
+from pyomo.solvers.plugins.solvers.gurobi_direct import gurobipy
+from operator import itemgetter
+from pyomo.opt import TerminationCondition as tc
+from pyomo.contrib.mindtpy.cut_generation import add_oa_cuts
+
+
+@SolverFactory.register(
+    'mindtpy.oa',
+    doc='MindtPy: Mixed-Integer Nonlinear Decomposition Toolbox in Pyomo')
+class MindtPy_OA_Solver(_MindtPyAlgorithm):
+    """
+    Decomposition solver for Mixed-Integer Nonlinear Programming (MINLP) problems.
+
+    The MindtPy (Mixed-Integer Nonlinear Decomposition Toolbox in Pyomo) solver 
+    applies a variety of decomposition-based approaches to solve Mixed-Integer 
+    Nonlinear Programming (MINLP) problems. 
+    These approaches include:
+
+    - Outer approximation (OA)
+    - Global outer approximation (GOA)
+    - Regularized outer approximation (ROA)
+    - LP/NLP based branch-and-bound (LP/NLP)
+    - Global LP/NLP based branch-and-bound (GLP/NLP)
+    - Regularized LP/NLP based branch-and-bound (RLP/NLP)
+    - Feasibility pump (FP)
+
+    This solver implementation has been developed by David Bernal <https://github.com/bernalde>
+    and Zedong Peng <https://github.com/ZedongPeng> as part of research efforts at the Grossmann
+    Research Group (http://egon.cheme.cmu.edu/) at the Department of Chemical Engineering at 
+    Carnegie Mellon University.
+    """
+    CONFIG = _get_MindtPy_OA_config()
+
+    def available(self, exception_flag=True):
+        """Check if solver is available.
+        """
+        return True
+
+    def license_is_valid(self):
+        return True
+
+    def version(self):
+        """Return a 3-tuple describing the solver version."""
+        return __version__
+
+    def solve(self, model, **kwds):
```

Comment:
Again, this is extremely similar if not identical to some of the other solve methods--it could draw from common functionality in a base class solve method.

Reply:
I think `solve` method should be defined alone in each algorithm, even if they look similar. Actually, they are just similar, not identical. We can define utility methods in the base class, however I don't `solve` is a utility method. So as `MindtPy_iteration_loop`.